pkg/cmd/server/start/start_node.go

```diff
@@ -255,22 +253,23 @@ func RunSDNController(config *kubernetes.NodeConfig, nodeConfig configapi.NodeCo
 	if err != nil {
 		glog.Fatal("Failed to get kube client for SDN")
 	}
-	registry := osdn.NewOsdnRegistryInterface(oclient, config.Client)
-
-	switch nodeConfig.NetworkConfig.NetworkPluginName {
-	case flatsdn.NetworkPluginName():
-		ch := make(chan struct{})
-		config.KubeletConfig.StartUpdates = ch
-		go flatsdn.Node(registry, nodeConfig.NodeName, nodeConfig.NodeIP, ch, nodeConfig.NetworkConfig.MTU)
-	case multitenant.NetworkPluginName():
-		ch := make(chan struct{})
-		config.KubeletConfig.StartUpdates = ch
-		plugin := multitenant.GetKubeNetworkPlugin()
-		config.KubeletConfig.NetworkPlugins = append(config.KubeletConfig.NetworkPlugins, plugin)
-		go multitenant.Node(registry, nodeConfig.NodeName, nodeConfig.NodeIP, ch, plugin, nodeConfig.NetworkConfig.MTU)
-		return registry
+
+	ch := make(chan struct{})
+	config.KubeletConfig.StartUpdates = ch
+
+	controller, err := factory.NewPlugin(nodeConfig.NetworkConfig.NetworkPluginName, oclient, config.Client, nodeConfig.NodeName, nodeConfig.NodeIP, ch)
+	if err != nil {
+		glog.Fatalf("SDN initialization failed: %v", err)
 	}
-	return nil
+
+	config.KubeletConfig.NetworkPlugins = append(config.KubeletConfig.NetworkPlugins, controller)
+
+	endpointFilter, err := controller.StartNode(nodeConfig.NetworkConfig.MTU)
```

Comment:
OK, this is the problem. The old code invoked the node startup as a goroutine, but the new code calls it directly, which won't work, because controller.StartNode() won't return until the node's subnet appears in the registry, but the code that causes that to happen doesn't run until after this function returns.

And since this function can't return until after it has gotten the endpointFilter from the plugin, there's no way to fix this without changing things in openshift-sdn again...